lib/cache_mngr.c

```diff
@@ -185,23 +220,13 @@ int nl_cache_mngr_alloc(struct nl_sock *sk, int protocol, int flags,
 	if ((err = nl_socket_set_nonblocking(mngr->cm_sock)) < 0)
 		goto errout;
 
-	/* Create and allocate socket for sync cache fills */
-	mngr->cm_sync_sock = nl_socket_alloc();
-	if (!mngr->cm_sync_sock) {
-		err = -NLE_NOMEM;
-		goto errout;
-	}
-	if ((err = nl_connect(mngr->cm_sync_sock, protocol)) < 0)
-		goto errout_free_sync_sock;
 
 	NL_DBG(1, "Allocated cache manager %p, protocol %d, %d caches\n",
 	       mngr, protocol, mngr->cm_nassocs);
 
 	*result = mngr;
 	return 0;
 
-errout_free_sync_sock:
-	nl_socket_free(mngr->cm_sync_sock);
```

Comment:
like `sk`, the new parameter `sync_sk`/`sk_blocking` should be optional/nullable. If the user did not specify a blocking socket, in case of error the internally allocated one will still be needed to released.

You need to right logic, so that all internally allocated data is correctly freed on every error path. Use `_nl_auto_nl_socket` instead of explicit free for that.